### PR TITLE
Stop SpiClient soundness from regressing

### LINKED ISSUE #1211: Add support for "compiletest" tests
Something that came out of PR #1210 is that there's some code patterns we'd like to ensure don't compile, which would be a way to help us enforce some borrowing rules/requirements around (at least) Spi.

That code is:

```rust
#[pg_test]
fn issue1209() -> Result<Option<String>, Box<dyn Error>> {
    // create the cursor we actually care about
    let mut res = Spi::connect(|c| {
        c.open_cursor("select 'hello' from generate_series(1, 10000)", None)
            .fetch(10000)
            .unwrap()
    });

    // here we just perform some allocations to make sure that the previous cursor gets invalidated
    for _ in 0..1000 {
        Spi::connect(|c| c.open_cursor("select 1", None).fetch(1).unwrap());
    }

    // later elements are probably more likely to point to deallocated memory
    for _ in 0..1000 {
        res.next();
    }

    // segfault
    Ok(res.next().unwrap().get::<String>(1)?)
}

#[pg_extern]
fn issue1209_prepared_stmt(q: &str) -> Result<Option<String>, Box<dyn std::error::Error>> {
    use pgrx::spi::Query;

    let prepared = { Spi::connect(|c| c.prepare(q, None))? };

    Ok(Spi::connect(|c| prepared.execute(&c, Some(1), None)?.first().get(1))?)
}
```

@workingjubilee suggested we use the `ui_test` crate (https://github.com/pgcentralfoundation/pgrx/pull/1210#issuecomment-1634730902).

### PR DESCRIPTION
**Pull request by workingjubilee**
...by adding a regression test with trybuild!
Adds ui tests using trybuild's test runner.

Fixes #1211.

### CODE REVIEW

File: pgrx-tests/tests/ui/escaping-spiclient-1209-cursor.rs [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+    // later elements are probably more likely to point to deallocated memory
+    for _ in 0..1000 {
+        res.next();
+    }
+
+    // segfault
+    Ok(res.next().unwrap().get::<String>(1)?)
+}
+
+fn main() {}
```
Comment (workingjubilee) [inline pgrx-tests/tests/ui/escaping-spiclient-1209-cursor.rs:27]: I don't like the frivolous mains but whatever.

**Comment by workingjubilee** [comment]
I got CI functional but this now needs cleanup in order to be ready to land again, it is not acceptable in its current state.

**Comment by workingjubilee** [comment]
I hate that this works.

File: pgrx-tests/tests/ui/escaping-spiclient-1209-cursor.rs (hunk 1 above)
Comment (thomcc) [reply pgrx-tests/tests/ui/escaping-spiclient-1209-cursor.rs:27]: might be able to avoid with `#![crate_type = "lib"]`? not sure.

**Review by thomcc — APPROVED** [review]
Seems fine. Worried about some of the big constants WRT #1218 but they shouldn't be *too* big...

File: pgrx-tests/tests/ui/escaping-spiclient-1209-cursor.rs (hunk 1 above)
Comment (workingjubilee) [reply pgrx-tests/tests/ui/escaping-spiclient-1209-cursor.rs:27]: Tried, unfortunately didn't work.

**Pull request merged**